File: PROYECTO ARBOL AVL/Propietario.cpp
#include "Propietario.h"
#include "Validaciones.h"
#include <algorithm>
#include "ValidacionPlaca.h"
#include "Lista.h"

using namespace std;

Propietario::Propietario()
    : nombre(""), apellido(""), cedula(""), correo(""), placas() {}

Propietario::Propietario(string nombre, string apellido, string cedula, string correo)
    : nombre(nombre), apellido(apellido), cedula(cedula), correo(correo) {}

Propietario::Propietario(string nombre, string apellido, string cedula, string correo, vector<string> placas)
    : nombre(nombre), apellido(apellido), cedula(cedula), correo(correo), placas(placas) {}


void Propietario::setNombre(const string &nombre)
{
    this->nombre = nombre;
}

void Propietario::setApellido(const string &apellido)
{
    this->apellido = apellido;
}

void Propietario::setCedula(const string &cedula)
{
    this->cedula = cedula;
}

void Propietario::setCorreo(const string &correo)
{
    this->correo = correo;
}

vector<string> Propietario::eliminarPlaca(const string &placa)
{

    auto it = find(placas.begin(), placas.end(), placa);
    if (it != placas.end()) {
        placas.erase(it);
        cout << "Placa " << placa << " eliminada correctamente." << endl;
    } else {
        cout << "La placa no se encontró en las placas del propietario." << endl;
    }

    return placas;
}

string Propietario::getNombre() const
{
    return nombre;
}

string Propietario::getApellido() const
{
    return apellido;
}

string Propietario::getCedula() const
{
    return cedula;
}


string Propietario::getCorreo() const
{
    return correo;
}
void Propietario::setPlacas(const vector<string> &placas)
{
    this->placas = placas;
}

void Propietario::agregarPlaca(const string &placa)
{
    placas.push_back(placa);
}

vector<string> Propietario::getPlacas() const
{
    return placas;
}

ostream &operator<<(ostream &os, const Propietario &propietario)
{
    os << "Nombre: " << propietario.getNombre() << "\nApellido: " << propietario.getApellido()
       << "\nCedula: " << propietario.getCedula() << "\nCorreo: " << propietario.getCorreo()
       << "\n----------------------------------------" << endl;

    os << "Placas: ";
    if (propietario.getPlacas().empty())
    {
        os << "N/A";
    }
    else
    {
        for (const auto &placa : propietario.getPlacas())
        {
            os << placa << " ";
        }
    }
    os << "\n----------------------------------------" << endl;

    return os;
}


Propietario Propietario::agregarPropietario(ListaCircularDoble<Propietario> &listaPropietarios)
{
    Validaciones validaciones;
    string nombre = validaciones.ingresarString("Ingrese el nombre del propietario: ");
    string apellido = validaciones.ingresarString("Ingrese el apellido del propietario: ");
    string cedula;

    do {
        cedula = validaciones.ingresarCedula("Ingrese la cedula del propietario: ");
        if (!validaciones.validarCedula(cedula)) {
            cout << "Cédula inválida. Intente de nuevo." << endl;
        }
    } while (!validaciones.validarCedula(cedula));

    Nodo<Propietario> *propietarioNodo = nullptr;
    bool propietarioEncontrado = false;

   
    while (true) {
        propietarioNodo = listaPropietarios.getPrimero();
        propietarioEncontrado = false;

        do {
            if (propietarioNodo->getDato().getCedula() == cedula) {
                propietarioEncontrado = true;
                break;
            }
            propietarioNodo = propietarioNodo->getSiguiente();
        } while (propietarioNodo != listaPropietarios.getPrimero());

        if (propietarioEncontrado) {
            cout << "Ya existe un propietario con esa cedula. Por favor ingrese una cedula diferente." << endl;
            
            cedula = validaciones.ingresarCedula("Ingrese el numero de cedula del propietario: ");
        } else {
            cout << "Cedula disponible. Puede continuar." << endl;
            break;  
        }
    }

  
    string correo;
    do {
        correo = validaciones.ingresarCorreo("Ingrese el correo electronico del propietario: ");
        if (!validaciones.validarCorreo(correo)) {
            cout << "Correo invalido. Intente de nuevo." << endl;
        }
    } while (!validaciones.validarCorreo(correo));

    
    Propietario nuevoPropietario(nombre, apellido, cedula, correo);

    cout << "Propietario creado exitosamente." << endl;

    return nuevoPropietario;
}

File: PROYECTO PYTHON/Propietario.py
from Validaciones import Validaciones


class Propietario:
    def __init__(self, nombre="", apellido="", cedula="", correo="", placas=None):
        """
        Constructor de la clase Propietario.
        Inicializa los atributos del propietario.
        """
        if placas is None:
            placas = []
        self.nombre = nombre
        self.apellido = apellido
        self.cedula = cedula
        self.correo = correo
        self.placas = placas

    def set_nombre(self, nombre):
        """
        Establece el nombre del propietario.
        """
        self.nombre = nombre

    def set_apellido(self, apellido):
        """
        Establece el apellido del propietario.
        """
        self.apellido = apellido

    def set_cedula(self, cedula):
        """
        Establece la cédula del propietario.
        """
        self.cedula = cedula

    def set_correo(self, correo):
        """
        Establece el correo electrónico del propietario.
        """
        self.correo = correo

    def eliminar_placa(self, placa):
        """
        Elimina una placa de la lista de placas del propietario.
        """
        if placa in self.placas:
            self.placas.remove(placa)
            print(f"Placa {placa} eliminada correctamente.")
        else:
            print("La placa no se encontró en las placas del propietario.")
        return self.placas

    def get_nombre(self):
        """
        Obtiene el nombre del propietario.
        """
        return self.nombre

    def get_apellido(self):
        """
        Obtiene el apellido del propietario.
        """
        return self.apellido

    def get_cedula(self):
        """
        Obtiene la cédula del propietario.
        """
        return self.cedula

    def get_correo(self):
        """
        Obtiene el correo electrónico del propietario.
        """
        return self.correo

    def set_placas(self, placas):
        """
        Establece la lista de placas del propietario.
        """
        self.placas = placas

    def agregar_placa(self, placa):
        """
        Agrega una placa a la lista de placas del propietario.
        """
        self.placas.append(placa)

    def get_placas(self):
        """
        Obtiene la lista de placas del propietario.
        """
        return self.placas

    def __str__(self):
        """
        Representación en cadena de texto del propietario.
        """
        propietario_info = (
            f"Nombre: {self.get_nombre()}\n"
            f"Apellido: {self.get_apellido()}\n"
            f"Cedula: {self.get_cedula()}\n"
            f"Correo: {self.get_correo()}\n"
            f"{'-'*40}\n"
        )
        placas_info = "Placas: "
        if not self.get_placas():
            placas_info += "N/A"
        else:
            placas_info += " ".join(self.get_placas())
        placas_info += f"\n{'-'*40}"
        return propietario_info + placas_info

    @staticmethod
    def agregar_propietario():
        """
        Método estático para crear un nuevo propietario.
        Solicita los datos del propietario y los valida.
        """
        validaciones = Validaciones()
        nombre = validaciones.ingresar_string("Ingrese el nombre del propietario: ")
        apellido = validaciones.ingresar_string("Ingrese el apellido del propietario: ")

        cedula = ""
        while not validaciones.validar_cedula(cedula):
            cedula = validaciones.ingresar_cedula("Ingrese la cedula del propietario: ")
            if not validaciones.validar_cedula(cedula):
                print("Cédula inválida. Intente de nuevo.")

        correo = ""
        while not validaciones.validar_correo(correo):
            correo = validaciones.ingresar_correo("Ingrese el correo electrónico del propietario: ")
            if not validaciones.validar_correo(correo):
                print("Correo inválido. Intente de nuevo.")

        nuevo_propietario = Propietario(nombre, apellido, cedula, correo)
        print("Propietario creado exitosamente.")
        return nuevo_propietario